### ISSUE TITLE
ProtoFlux feedback value field drive evaluates twice per update

### ISSUE THREAD

**Issue by Nytra** [Yellow-Dog-Man/Resonite-Issues#1110]
### Describe the bug?

This is kind of a continuation / spinoff of https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1071

It might be caused by the same underlying problem.

I noticed that some ProtoFlux drives appeared to change the value too quickly in some circumstances.

This replication uses the same setup as in that issue, but with an added Update node computing the Delta of the value in the ValueField.

### To Reproduce

1) Spawn the item in this public folder: resrec:///U-Nytra/R-BB57A009409B0C329250E06B173779BA13EC4B036E03456A3C411BCBAE894226

2) Unpack the ProtoFlux in the spawned slot.

3) Observe that with the Continuous Relay in place the value stored in the DataModelStore says 0.1. This is the Delta of value in the ValueField. 0.1 is the correct value that it should ideally be.

4) Disconnect the Continuous Relay and connect the Add node directly into the Drive node.

5) The value in the ValueField stops updating (This is https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/1071)

6) Follow the steps in the linked issue above and change the value in the ValueInput node to make the Drive node start working again.

7) Observe that the value in the DataModelStore is double of what it should be now. It is double of the value in the ValueInput. Which means that that value is being added twice per update which shouldn't happen.

### Expected behavior

The drive evaluates once per update.

### Screenshots

![2024-01-09 21 09 33](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/14206961/1ab1685d-26f0-4a07-9234-2e831d9cf52c)

![2024-01-09 21 09 13](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/14206961/e6abe7d6-64c7-4fdc-a0e5-2ad24356f6cf)

### Resonite Version Number

Beta 2024.1.3.1265

### What Platforms does this occur on?

Windows

### What headset if any do you use?

Valve Index, Quest 2, Desktop

### Log Files

[DESKTOP-H976HO2 - 2024.1.3.1265 - 2024-01-09 21_05_24.log](https://github.com/Yellow-Dog-Man/Resonite-Issues/files/13879833/DESKTOP-H976HO2.-.2024.1.3.1265.-.2024-01-09.21_05_24.log)

### Additional Context

N/a

### Reporters

Nytra (Discord: nytra)

**Comment by Frooxius** [comment]
I don't think this is technically a bug. We do not guarantee that drives only get evaluated once per frame, so this is technically valid behavior.

We can still look at it, though there's a chance we might not fix this. What I suspect that happens is that because the drive causes change to occur, this schedules the drive to be updated again, in the second phase of changes - ones that were caused by changes.

As such, adding a system that specifically checks when the change happens comes from a feedback loop might end up adding too much complexity. We will see when we actually look into this.

However my generally recommendation would be not to expect drives to be evaluated certain amounts of times per frame - generally the value you're driving for should be valid at any point when it gets evaluated. If you need specific control over timings, using impulses is recommended.

**Comment by Nytra** [comment]
Thanks for clarification.

It may be technically valid, however it is strange that it behaves differently with and without the Continuous Relay. I wonder why it isn't consistent.

**Comment by Frooxius** [comment]
It might be good to investigate. Normally drives will use change tracking, while continuous relay will use different system to schedule the updates to happen.

**Comment by Banane9** [comment]
> I don't think this is technically a bug. We do not guarantee that drives only get evaluated once per frame, so this is technically valid behavior.
> 
> We can still look at it, though there's a chance we might not fix this.

I'm not sure whether this was also the case on the old platform, but I do know that some legacy content makes use of these drive loops - so making them only evaluate once would be desirable if it's not too much hassle.

**Comment by lxw404** [comment]
This does break my understanding of building feedback systems, which I thought would only have handled singular updates.  I have built some PID controllers as a feedback loop similar to what was shown here and if they're accumulating multiple times, then that could lead to chaotic feedback, or undesired behavior...
